scanpy/ann_data.py

```diff
@@ -164,27 +202,37 @@ def __init__(self, X, smp=None, var=None, vis=None, **meta):
         self.vis = vis or {}
         self._meta = meta
 
-    @classmethod
-    def from_ddata(cls, X=None, rownames=None, colnames=None, **meta):
-        """
-        Temporary helper to an AnnData from a “ddata” dict.
-
-        Parameters
-        ----------
-        X : n×p data matrix
-        rownames : n-length array of samples
-        colnames : n-length array of variables
-        """
-        if X is None:
-            raise ValueError('Missing data. Got instead {}'.format(meta.keys()))
-
-        smp_meta = var_meta = None
-        if rownames is not None:
-            smp_meta = np.rec.fromarrays([np.asarray(rownames)], names=[SMP_NAMES])
-        if colnames is not None:
-            var_meta = np.rec.fromarrays([np.asarray(colnames)], names=[VAR_NAMES])
-
-        return cls(X, smp_meta, var_meta, **meta)
+        if ddata is not None:
+            for key, value in ddata.items():
+                if not key in ['smp', 'row', 'var', 'col']:
+                    self._meta[key] = value
+                elif key == 'row':
+                    for k, v in ddata['row'].items():
+                        self.smp[k] = v
+                elif key == 'smp':
+                    for k, v in ddata['smp'].items():
+                        self.smp[k] = v
+                elif key == 'col':
+                    for k, v in ddata['col'].items():
+                        self.var[k] = v
+                elif key == 'var':
+                    for k, v in ddata['var'].items():
+                        self.var[k] = v
+                    
+    def smp_keys(self):
+        return list(self.smp.dtype.names)[1:]
```

Comment:
yes, that's a hack. but I think the np.recarray is ordered, right?

both your suggestions are fine. i just wanted it to get to work. I'm happy if you go over it again. maybe not right now, as everything is tested. but we can sit together maybe tomorrow, if you want.